### PR TITLE
Add support for macOS

### PR DESCRIPTION
**Pull request by kalbasit**
This pull request adds support for macOS.

I had to move the `mode` declaration from being computed in Soxin to being manually assigned in SoxinCFG. This allows us to build a system of any kind by simply set the `mode` input to a specific value.

**Open questions**

- Is the `mode` the best name for this input?

SoxinCFG: https://github.com/kalbasit/soxincfg/pull/20

### CODE REVIEW

File: lib/system-flake.nix [hunk 1]
```diff
@@ -72,43 +89,192 @@ let
...
+                # include sane flake defaults from flake-utils-plus which sets sane `nix.*` defaults.
+                # Please refer to implementation/readme in
+                # github:gytis-ivaskevicius/flake-utils-plus for more details.
+                ++ (singleton flake-utils-plus.nixosModules.saneFlakeDefaults)
+                # include the NixOS modules
+                ++ extraNixosModules
+                # include Soxin modules
+                ++ (singleton soxin.nixosModule)
+                # include home-manager modules
+                ++ (singleton home-manager.nixosModules.home-manager)
```
Comment (rissson) [inline lib/system-flake.nix:235]: I wonder if there is a way to factorize all of that code.
Somewhere along the lines of:
```
commonModulesBases = {
  sops-nix = "sops";
  flake-utils-plus = "saneFlakeDefaults";
  extraModules = "modules";
  /* ... */
};

commonModulesDarwin = mapAttrsToList (name: value: inputs.${name}.darwinModules.${value});
commonModulesNixOS = mapAttrsToList (name: value: inputs.${name}.nixosModules.${value});
```

what do you think?

File: modules/programs/tmux/default.nix [hunk 2]
```diff
@@ -99,7 +99,9 @@ in
 
   config = mkIf cfg.enable (mkMerge [
     # add all plugins installed by themes
-    { soxin.programs.tmux.plugins = cfg.theme.plugins; }
+    (optionalAttrs (mode == "NixOS" || mode == "home-manager") {
+      soxin.programs.tmux.plugins = cfg.theme.plugins;
```
Comment (rissson) [inline modules/programs/tmux/default.nix:103]: Nothing uses `tmux.plugins` in Darwin if I'm not mistaken. Why was this needed?

File: lib/system-flake.nix (hunk 1 above)
Comment (kalbasit) [reply lib/system-flake.nix:235]: Yes, I agree but thought of doing a refactor later as it's best to get the functionality working correctly first.

File: modules/programs/tmux/default.nix (hunk 2 above)
Comment (kalbasit) [reply modules/programs/tmux/default.nix:103]: Not darwin but home-manager on darwin. My tmux works fine on Darwin right now, themes and all. Without this change, this was being set for nix-darwin that does not have the option.

**Comment by rissson** [comment]
>  Is the `mode` the best name for this input?

Maybe `backend` is better?

File: template/flake.nix [hunk 3]
```diff
@@ -54,6 +59,7 @@
       channelsConfig = {
         # allowBroken = true;
         allowUnfree = true;
+        # allowUnsupportedSystem = true;
```
Comment (rissson) [inline template/flake.nix:62]: Was this needed for M1?

File: lib/mk-flake.nix [hunk 4]
```diff
@@ -109,33 +118,113 @@ let
...
+                # include the NixOS special arguments.
+                // nixosSpecialArgs
+              ;
+            }
+          ))
+          nixosOnlyHosts;
+
+      # TODO: add support for home-manager modes
+    in
+    darwinHosts // nixosHosts;
```
[the remark this answers is not in the document]
Comment (kalbasit) [reply lib/mk-flake.nix:219]: I think there's room for improvement, but out of scope here. Perhaps we can revisit this, and the `mode` naming together in a new design.

File: lib/mk-flake.nix [hunk 5]
```diff
@@ -235,22 +324,17 @@ let
...
             ++ (singleton { nix = { inherit generateNixPathFromInputs generateRegistryFromInputs linkInputs; }; })
             # configure home-manager
             ++ (singleton {
-            # tell home-manager to use the global (as in NixOS system-level) pkgs and
-            # install all  user packages through the users.users.<name>.packages.
+            # tell home-manager to use the global (as in NixOS/Nix-Darwin
+            # system-level) pkgs and install all user packages through the
+            # users.users.<name>.packages.
             home-manager.useGlobalPkgs = true;
             home-manager.useUserPackages = true;
```
[the remark this answers is not in the document]
Comment (kalbasit) [reply lib/mk-flake.nix:339]: True, but out of scope for this PR. It should be its own change. Can you submit a separate PR once these are merged down?

File: template/flake.nix (hunk 3 above)
Comment (kalbasit) [reply template/flake.nix:62]: The unsupported system? I don't remember actually why I added this, maybe when I was testing M1 back then. Either way, it shouldn't matter since it's commented but I can remove it if you think it can confuse people.

File: template/flake.nix [hunk 6]
```diff
@@ -87,5 +93,8 @@
 
       # include all overlays
       overlay = import ./overlays;
+
+      # set the nixos specialArgs
+      nixosSpecialArgs = { inherit nixos-hardware; };
```
[the remark this answers is not in the document]
Comment (kalbasit) [reply template/flake.nix:98]: yes, we don't currently do that, and I was not able to access inputs the other day. Out of scope though.

File: lib/mk-flake.nix (hunk 4 above)
[the remark this answers is not in the document]
Comment (rissson) [reply lib/mk-flake.nix:219]: Fair enough

File: lib/mk-flake.nix (hunk 5 above)
[the remark this answers is not in the document]
Comment (rissson) [reply lib/mk-flake.nix:339]: I created https://github.com/SoxinOS/soxin/issues/55

File: template/flake.nix (hunk 3 above)
Comment (rissson) [reply template/flake.nix:62]: I don't think it matters that much.

File: template/flake.nix (hunk 6 above)
[the remark this answers is not in the document]
Comment (rissson) [reply template/flake.nix:98]: Indeed.

**Review by rissson — APPROVED** [review]

**Comment by rissson** [comment]
Did you have a look at my take on this in `risson-darwin`?
